### PR TITLE
Fix/ensure wsg deletion

### PR DESCRIPTION
**Pull request by csmartins**
## Context

When the clusters/igs are deleted, we must also clean up the resources that reference those objects like security groups. SGs have a list of referenced infrastructure resources which the reconciliation loop needs to ensure is up to date with the actual existing resources. In the current implementation if we delete a Machine pool there's a race condition which can cause the reconciliation to fail in the Detach process. We can also have the condition where the WSG resource is deleted but the SG still exists in AWS.

## Solution

To ensure the proper order of reconciliation and deletion we are leveraging the features of Finalizers. Now the infrastructure referenced by the WSG gets a finalizer, which we use to proper coordinate the deletions. Also, we are using the Status property of the WSG to guarantee that we properly detach the SG from a Machine Pool's ASG getting deleted.

### CODE REVIEW

File: internal/controller/ec2.aws/securitygroup_controller.go [hunk 1]
```diff
@@ -289,60 +322,18 @@ func (r *SecurityGroupReconciliation) reconcileNormal(ctx context.Context) (ctrl
...
-		default:
-			return resultError, fmt.Errorf("infrastructureRef not supported")
-		}
+	result, err = r.ensureDetachRemovedReferences(ctx, csg)
+	if err != nil {
+		return result, err
 	}
 
 	r.sg.Status.Ready = true
+	r.sg.Status.AppliedInfrastructureRef = r.sg.DeepCopy().Spec.InfrastructureRef
```
Comment (rafatio) [inline internal/controller/ec2.aws/securitygroup_controller.go]: why do we need to call this DeepCopy method here

File: internal/controller/ec2.aws/securitygroup_controller.go [hunk 2]
```diff
@@ -161,86 +179,101 @@ func (r *SecurityGroupReconciliation) retrieveInfraRefInfo(ctx context.Context)
...
+			if err := r.Client.Get(ctx, key, kcp); err != nil {
+				retrieveErr = err
+				continue
+			}
 
-		region, err := kopsutils.GetRegionFromKopsControlPlane(ctx, kcp)
-		if err != nil {
-			return fmt.Errorf("error retrieving region: %w", err)
-		}
+			region, err := kopsutils.GetRegionFromKopsControlPlane(ctx, kcp)
```
Comment (rafatio) [inline internal/controller/ec2.aws/securitygroup_controller.go]: this part is quite similar to the one above, the only difference between them is the retrieval of the kmp. WDYT about encapsulating what is similar between them?

File: internal/controller/ec2.aws/securitygroup_controller.go [hunk 3]
```diff
@@ -570,6 +561,7 @@ func (r *SecurityGroupReconciliation) reconcileDelete(ctx context.Context, sg *s
 	csg := &crossec2v1beta1.SecurityGroup{}
 	err := r.Get(ctx, key, csg)
 	if apierrors.IsNotFound(err) {
+		r.log.Info(fmt.Sprintf("CrossplaneSecurityGroup %v not found at %v, moving on with the deletion process", key.Name, key.Namespace))
```
Comment (rafatio) [inline internal/controller/ec2.aws/securitygroup_controller.go]: for the sake of consistency, let's start the logs in lower case

File: internal/controller/ec2.aws/securitygroup_controller.go [hunk 4]
```diff
@@ -585,33 +577,53 @@ func (r *SecurityGroupReconciliation) reconcileDelete(ctx context.Context, sg *s
 				Name:      infrastructureRef.Name,
 				Namespace: infrastructureRef.Namespace,
 			}
-			if err := r.Client.Get(ctx, key, &kmp); err != nil {
-				return resultError, err
+			err := r.Client.Get(ctx, key, &kmp)
+			if apierrors.IsNotFound(err) {
+				r.log.Error(err, fmt.Sprintf("KopsMachinePool %v not found at %v, moving on with the deletion process", key.Name, key.Namespace))
```
Comment (rafatio) [inline internal/controller/ec2.aws/securitygroup_controller.go]: for the sake of consistency, let's start the logs in lower case

File: internal/controller/ec2.aws/securitygroup_controller.go [hunk 5]
```diff
@@ -585,33 +577,53 @@ func (r *SecurityGroupReconciliation) reconcileDelete(ctx context.Context, sg *s
...
+			kcp := kcontrolplanev1alpha1.KopsControlPlane{}
 			key := client.ObjectKey{
 				Namespace: infrastructureRef.Namespace,
 				Name:      infrastructureRef.Name,
 			}
-			if err := r.Client.Get(ctx, key, kcp); err != nil {
-				return resultError, err
+			err := r.Client.Get(ctx, key, &kcp)
+			if apierrors.IsNotFound(err) {
+				r.log.Error(err, fmt.Sprintf("KopsControlPlane %v not found at %v, moving on with the deletion process", key.Name, key.Namespace))
```
Comment (rafatio) [inline internal/controller/ec2.aws/securitygroup_controller.go]: for the sake of consistency, let's start the logs in lower case

File: internal/controller/ec2.aws/securitygroup_controller.go (hunk 1 above)
Comment (csmartins) [reply internal/controller/ec2.aws/securitygroup_controller.go]: No specific reason, I guess we saw that in another place and just copied without properly judging the why

Comment (csmartins) [reply internal/controller/ec2.aws/securitygroup_controller.go]: But we can just assign the value directly

File: internal/controller/ec2.aws/securitygroup_controller.go (hunk 2 above)
Comment (csmartins) [reply internal/controller/ec2.aws/securitygroup_controller.go]: makes sense, this could be a private function

Comment (csmartins) [reply internal/controller/ec2.aws/securitygroup_controller.go]: fixed in the latest changes, but only extracted the repeated kcp part that was repeated

**Review by rafatio — APPROVED** [review]

**Pull request merged**